### PR TITLE
Improve arguments object

### PR DESCRIPTION
**Pull request by rerobika**
- Enhancement: Arguments object properties are now lazy instantiated
 - Bugfix: Mapped arguments object instantiated properties cannot be lcached
 - Bugfix: Mapped arguments should be constructed even if 0 formal parameters or arguments are provided
 - Update: remove 'caller' property of unmapped arguments object

JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/base/ecma-gc.c [hunk 1]
```diff
@@ -148,6 +148,36 @@ ecma_deref_object (ecma_object_t *object_p) /**< object */
...
 
+/**
+ * Mark objects referenced by arguments object
+ */
+static void
+ecma_gc_mark_arguments_object (ecma_extended_object_t *ext_object_p) /**< arguments object */
+{
+  JERRY_ASSERT (ecma_get_object_type ((ecma_object_t *) ext_object_p) == ECMA_OBJECT_TYPE_PSEUDO_ARRAY);
+
+  ecma_unmapped_arguments_t *arguments_p = (ecma_unmapped_arguments_t *) ext_object_p;
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-gc.c:159]: ecma_unmapped_arguments_t?

File: jerry-core/ecma/base/ecma-gc.c [hunk 2]
```diff
@@ -880,6 +907,39 @@ ecma_gc_free_native_pointer (ecma_property_t *property_p) /**< property */
...
+  }
+
+  ecma_value_t *argv_p = (ecma_value_t *) (((uint8_t *) ext_object_p) + object_size);
+  ecma_unmapped_arguments_t *arguments_p = (ecma_unmapped_arguments_t *) ext_object_p;
+  uint32_t arguments_number = arguments_p->header.u.pseudo_array.u2.arguments_number;
+
+  for (uint32_t i = 0; i < arguments_number; i++)
+  {
+    ecma_free_value_if_not_object (argv_p[i]);
+  }
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-gc.c:938]: What happens if new items are appended to the arguments object?

File: jerry-core/ecma/base/ecma-globals.h [hunk 3]
```diff
@@ -2148,6 +2148,41 @@ typedef struct
...
+  ECMA_ARGUMENTS_OBJECT_NO_FLAGS = 0,                    /* unmapped arguments object */
+  ECMA_ARGUMENTS_OBJECT_MAPPED = (1 << 0),               /* mapped arguments object */
+  ECMA_ARGUMENTS_OBJECT_CALLEE_INITIALIZED = (1 << 1),   /* 'callee' property has been lazy initialized */
+  ECMA_ARGUMENTS_OBJECT_CALLER_INITIALIZED = (1 << 2),   /* 'caller' property has been lazy initialized */
+  ECMA_ARGUMENTS_OBJECT_LENGTH_INITIALIZED = (1 << 3),   /* 'length' property has been lazy initialized */
+  ECMA_ARGUMENTS_OBJECT_ITERATOR_INITIALIZED = (1 << 4), /* 'Symbol.iterator' property has been lazy initialized */
+} ecma_arguments_object_flags_t;
+
+/**
+ * Definition of unmapped arguments object
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-globals.h:2165]: Sometimes there is a dot after sentences, sometimes not. Please make it consistent.

File: jerry-core/ecma/base/ecma-globals.h [hunk 4]
```diff
@@ -2148,6 +2148,41 @@ typedef struct
...
+typedef struct
+{
+  ecma_extended_object_t header; /**< object header */
+  ecma_value_t callee; /**< 'callee' property */
+} ecma_unmapped_arguments_t;
+
+/**
+ * Definition of mapped arguments object.
+ *
+ * Note: this structure must be kept in sync with `ecma_unmapped_arguments_t`
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-globals.h:2176]: Can we use ecma_unmapped_arguments_t as `header`?

File: jerry-core/ecma/operations/ecma-objects.c [hunk 5]
```diff
@@ -293,43 +293,59 @@ ecma_op_object_get_own_property (ecma_object_t *object_p, /**< the object */
     {
       property_p = ecma_op_bound_function_try_to_lazy_instantiate_property (object_p, property_name_p);
     }
+    else if (type == ECMA_OBJECT_TYPE_PSEUDO_ARRAY
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-objects.c:296]: There are many type checks here, switch is faster in this case.

File: jerry-core/ecma/operations/ecma-objects.c [hunk 6]
```diff
@@ -476,26 +492,23 @@ ecma_op_object_find_own (ecma_value_t base_value, /**< base value */
     {
       ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) object_p;
 
-      if (ext_object_p->u.pseudo_array.type == ECMA_PSEUDO_ARRAY_ARGUMENTS)
+      if (ext_object_p->u.pseudo_array.type == ECMA_PSEUDO_ARRAY_ARGUMENTS
+          && ext_object_p->u.pseudo_array.extra_info & ECMA_ARGUMENTS_OBJECT_MAPPED)
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-objects.c:496]: I would put a bracket around the bitwise and operation.

File: jerry-core/ecma/base/ecma-gc.c (hunk 2 above)
Comment (rerobika) [reply jerry-core/ecma/base/ecma-gc.c:945]: These are freed by the object's property list.

File: tests/test262-esnext-excludelist.xml [hunk 7]
```diff
@@ -2927,23 +2926,66 @@
...
-  <test id="language/arguments-object/mapped/nonconfigurable-descriptors-with-param-assign.js"><reason></reason></test>
-  <test id="language/arguments-object/mapped/nonconfigurable-nonenumerable-nonwritable-descriptors-set-by-arguments.js"><reason></reason></test>
-  <test id="language/arguments-object/mapped/nonconfigurable-nonenumerable-nonwritable-descriptors-set-by-param.js"><reason></reason></test>
-  <test id="language/arguments-object/mapped/nonconfigurable-nonwritable-descriptors-set-by-arguments.js"><reason></reason></test>
-  <test id="language/arguments-object/mapped/nonconfigurable-nonwritable-descriptors-set-by-param.js"><reason></reason></test>
+  <test id="language/arguments-object/cls-decl-async-private-gen-meth-args-trailing-comma-multiple.js"><reason></reason></test>
+  <test id="language/arguments-object/cls-decl-async-private-gen-meth-args-trailing-comma-null.js"><reason></reason></test>
+  <test id="language/arguments-object/cls-decl-async-private-gen-meth-args-trailing-comma-single-args.js"><reason></reason></test>
+  <test id="language/arguments-object/cls-decl-async-private-gen-meth-args-trailing-comma-spread-operator.js"><reason></reason></test>
```
Comment (ossy-szeged) [inline tests/test262-esnext-excludelist.xml:2932]: These tests are already in the exclude list in the "ESNext stage 3 proposals: class fields, private methods and static class features" block at the end.

**Review by zherczeg — APPROVED** [review]
LGTM

**Review by galpeter — APPROVED** [review]
lgtm

**Pull request merged**